rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_wrapper.cpp

```diff
@@ -76,6 +77,79 @@ SqliteWrapper::~SqliteWrapper()
   }
 }
 
+void SqliteWrapper::apply_pragma_settings(
+  std::vector<std::string> & pragmas,
+  rosbag2_storage::storage_interfaces::IOFlag io_flag)
+{
+  // sqlite pragmas are assigned with either '= value' or '(value)' syntax, depending on pragma
+  const std::string pragma_assign = "=";
+  const std::string pragma_bracket = "(";
+
+  // Apply default pragmas if not overridden by user setting
+  {
+    // Used to check whether db is readable
+    const std::string schema = "schema_version";
+
+    typedef std::unordered_map<std::string, std::string> pragmas_map_t;
+    pragmas_map_t default_pragmas = {
+      {schema, ""}
+    };
+    if (io_flag == rosbag2_storage::storage_interfaces::IOFlag::READ_WRITE) {
+      const pragmas_map_t write_default_pragmas = {
+        {"journal_mode", pragma_assign + "WAL"},
+        {"synchronous", pragma_assign + "NORMAL"}
+      };
+      default_pragmas.insert(write_default_pragmas.begin(), write_default_pragmas.end());
+    }
+    for (const auto & kv : default_pragmas) {
+      auto key = kv.first;
+      auto default_assignment = kv.second;
+      auto pragma_in_settings = std::find_if(
+        pragmas.begin(), pragmas.end(),
+        [key](const auto & pragma) {
+          return pragma.rfind(key, 0) == 0;
+        });
+      if (pragma_in_settings == pragmas.end()) {
+        pragmas.push_back(key + default_assignment);
+      }
+    }
+  }
+
+  for (auto & pragma : pragmas) {
+    if (pragma.empty()) {
+      continue;
+    }
+
+    // Extract pragma name
+    auto pragma_name = pragma;
+    auto found_value_assignment = pragma.find(pragma_assign);
+    if (found_value_assignment == std::string::npos) {
+      found_value_assignment = pragma.find(pragma_bracket);
+    }
+    if (found_value_assignment != std::string::npos) {
+      if (found_value_assignment == 0) {
+        // Incorrect syntax, starts with = or (
+        std::stringstream errmsg;
+        errmsg << "Incorrect storage setting syntax: " << pragma;
+        throw SqliteException{errmsg.str()};
+      }
+      // Strip value assignment part, trim trailing whitespaces before = or (
+      pragma_name = pragma.substr(0, found_value_assignment);
+      const std::string trailing_set(" \t\f\v\n\r");
+      pragma_name = pragma_name.substr(0, pragma_name.find_last_not_of(trailing_set) + 1);
```

Comment:
again I think this should happen in the `parse_pragmas` function, rather than the application here.

Reply:
Here, besides some double checking (which could indeed happen in parser), we are also creating a read-value statement from a set-value one (`PRAGMA name; `from `PRAGMA name = VALUE;`).

We could parse pragma statements into name and value in the parser already, I could move the code there.

Reply:
This should be fixed with the last commit